### PR TITLE
ScriptAPI - Basic inter-script communication! As seen in #22!

### PR DESCRIPTION
**Pull request by ashquarky**
# The ScriptAPI

Here it is! The first version of the Dragonet ScriptAPI, as seen in issue #22!

**Important!** Due to changes required the base Script class, all DAPISes need to provide their own unique identifier (or UID) - generally the plugin's name.
**All scripts will now need to include the following snippet:**

``` javascript
var UID = "Insert your UID here";
function getUID() {
    return UID;
}
```

And that's it! Now, let's get down to new features.
### Contacting other scripts

After creating your ScriptAPI object (ours will be called "Scripts") and your UID (stored in the "UID" variable) you can contact other scripts with `callCustomFunction()`. Here's how it works:

**Your script (script A)**

``` javascript
function Tick() {
    Scripts.callCustomFunction("sayHello");
}
```

**Another script (script B)**

``` javascript
function sayHello() {
    Server.clientMessage("Okay, I'll say hello!")
}
```

Now, every tick, script A will tell script B to run `sayHello()`, which will echo "Okay, I'll say hello!" into the chat.

The usefulness of this method of doing things is that if script A doesn't exist, script B will still run normally (minus the "sayHello" functionality, of course). Additionally, multiple scripts can call the same function, _and_ at the moment this can be used to call system functions (`callCustomFunction("Tick")`).

~~Okay, I've run out of time, I'll write up the info on the other method of contact tomorrow. Bear with me!~~
**_Update:**_ Okay, got it all finished. Thanks for waiting!
### Having other scripts contact you

Say, instead of telling a script when something happens to _you_, what happens if you want a script to tell you when something happens to _them_? Sure, you could have the other script run a custom function, but other scripts can hijack that and get called as well. What if you need to be the only one that knows? This is where custom methods come in. Here's an example:

**Your script (script A)**

``` javascript
function postInit() { //This is called AFTER Dragonet has finished loading, method assignments MUST be here
    Scripts.addMethod("addAlert", "addAlert_handle", UID); //You MUST include your UID variable
    //Explanation: ("name of method OTHER scripts will call", "what to actually call when this happens", UID);
}
function addAlert_handle() {
    //This will be called when script B triggers it: every tick
    Server.clientMessage("Alert Added");
}
```

**Another script (script B)**

``` javascript
function Tick() {
    Scripts.callCustomMethod("addAlert")
}
```

Simply put, every tick script B will tell script A to add an alert, and script A will have the handler method ("addAlert_handle") called when this happens. This is basically the first method in reverse. Neat, right?

**Quick note:** If 2 scripts try to register the same method at the same time, the one whose filename is first in alphabetical order succeds while the other one throws an error. This is because of the way Dragonet detects its scripts (cycling through the folder). In future we might block both and throw more annoying errors until the user fixes this.

### CODE REVIEW

**Comment by wordandahalf** [comment]
I'm going to merge the PR now (And not completely screw it up like last time)

**Comment by Creeperface01** [comment]
Good :)

**Comment by ashquarky** [comment]
@TheMCPEGamer Awesome. Do you want me to write up a wiki page or something or is it all good?

Also, any ideas for things to add (other than the whole custom function/method thing)

**Comment by Creeperface01** [comment]
Can you add method like useItem(x,y,z,itemId,blockId,side);

**Comment by wordandahalf** [comment]
@QuarkTheAwesome Yea, that'll be helpful if you make a new wiki page (I'll add it to the website's wiki later).

**Comment by DefinitlyEvil** [comment]
Well done guys!

**Comment by ashquarky** [comment]
@TheMCPEGamer Okay... So, like in the GitHub wiki?

**Comment by wordandahalf** [comment]
@QuarkTheAwesome  yes

**Comment by wordandahalf** [comment]
@QuarkTheAwesome Maybe we could change the UID to the Script's name? It makes more sense IMHO.

**Comment by ashquarky** [comment]
@TheMCPEGamer I went for UIDs because the name implied that they _must_ be unique. Anyway, you can change it if you want.

**Comment by ashquarky** [comment]
@TheMCPEGamer Okay, finished the wiki page!

Oh, and I think we can close the original issue (#22) with no issues.

**Comment by optrusty** [comment]
lol Why am I subbed to dragonet

**Comment by wordandahalf** [comment]
@QuarkTheAwesome  Thanks alot! :D
